pom.xml

```diff
@@ -118,6 +118,7 @@
             <groupId>net.i2p.crypto</groupId>
             <artifactId>eddsa</artifactId>
             <version>${eddsa.version}</version>
+            <scope>provided</scope>
```

Comment:
Who should provide it? As far as I understand, it is not part of the FIPS security providers. What will happen to applications deployed in non-FIPS environments when they upgrade to this version? Will it fail to find the dependency?

Reply:
Correct this is not a FIPS provider. But this is causing fips testcase failure because its get added into the security provider list. I have no idea why this security provider is there in the first place. Please let me know if you have any insight. 
But without this(provided scope) the munit test cases are passing. Does it sufficient? If not then we have to add this into the security provider of fips validation pipeline to pass the fips testcase.
Let me know your opinition.

Reply:
Hello, @dipeshkumardutta-sfemu ,
While analyzing the EdDSA issue, I discovered that the problem seems to originate from [here](https://github.com/apache/mina-sshd/blob/af6c8eddd897e18e8834c6439c48513cec150c8c/sshd-common/src/main/java/org/apache/sshd/common/util/security/SecurityUtils.java#L136). By default, the Apache SSHD Security Utils project is registering two security providers: BC and EdDSA. The good news is that there appears to be a mechanism to prevent this provider from being registered. Please take a look at the [SecurityUtils.setAPrioriDisabledProvider()](https://github.com/apache/mina-sshd/blob/af6c8eddd897e18e8834c6439c48513cec150c8c/sshd-common/src/main/java/org/apache/sshd/common/util/security/SecurityUtils.java#L209) method and its JavaDoc. We might be able to use this method to selectively avoid registering EdDSA, depending on whether we are running in a FIPS-compliant environment.

Reply:
Additionally, we need to test how the connector functions in a FIPS-compliant environment, particularly in terms of validating host keys. To my knowledge, this validation is typically handled by the EdDSA security provider. We must determine what happens to this feature when we opt not to register that provider. Could you please check if there are any test suites that specifically assess this validation?

Reply:
FYI Check [this](https://apache.googlesource.com/mina-sshd/+/ffe7a83e0890bafae43e41791c4ae629e2abda66/README.md#implemented_available-support):
```
Signatures/Keys: ssh-dss, ssh-rsa, rsa-sha2-256, rsa-sha2-512, nistp256, nistp384, nistp521 , 
ssh-ed25519 (requires eddsa optional module), 
sk-ecdsa-sha2-nistp256@openssh.com, sk-ssh-ed25519@openssh.com , ssh-rsa-cert-v01@openssh.com, ssh-dss-cert-v01@openssh.com, ssh-ed25519-cert-v01@openssh.com , ecdsa-sha2-nistp256-cert-v01@openssh.com, ecdsa-sha2-nistp384-cert-v01@openssh.com, ecdsa-sha2-nistp521-cert-v01@openssh.com
```
Is ssh-ed25519 the only one we would stop supporting in the case of not registering the EdDSA security provider?

Reply:
@acurci-at-mulesoft Yes as far as I get the information from internet, for host key verification EdDSA is used and in apache mina two security provider registerar is used BC and EdDSA. Looking through https://github.com/apache/mina-sshd/blob/e021a6f4597cb7259162169851bb9e4cb85e1a39/sshd-common/src/main/java/org/apache/sshd/common/util/security/eddsa/EdDSASecurityProviderRegistrar.java you see that it expects the class net.i2p.crypto.eddsa.EdDSAKey to exist:
So it means that it must have et.i2p.crypto.eddsa.EdDSAKe to exists.
As you mentioned I ran the munit test,  disabling the EdDSA in fips and BC in non-fips mode. The test execution was successful.
`if (FIPS_140_2_MODEL.equals(System.getProperty(MULE_SECURITY_MODEL))) {
      LOGGER.info("Running in fips mode so disabling EdDSA provider");
      SecurityUtils.setAPrioriDisabledProvider(SecurityUtils.EDDSA, true);
    } else {
      LOGGER.info("Running in non-fips mode so disabling BC provider");
      SecurityUtils.setAPrioriDisabledProvider(SecurityUtils.BOUNCY_CASTLE, true);
    }`
    
Also  have seen the host verification code in connector is shown below. Does it means that the Server key verification is ignored?
   `client
          .setServerKeyVerifier(new KnownHostsServerKeyVerifier(RejectAllServerKeyVerifier.INSTANCE, Paths.get(knownHostsFile)) {

            @Override
            protected boolean acceptKnownHostEntries(
                                                     ClientSession clientSession, SocketAddress remoteAddress,
                                                     PublicKey serverKey,
                                                     Collection<HostEntryPair> knownHosts) {
              if (GenericUtils.isEmpty(knownHosts)) {
                LOGGER.error("known_hosts collection is empty!");
              }
              return super.acceptKnownHostEntries(clientSession, remoteAddress, serverKey, knownHosts);
            }
          });`

Reply:
Host Key Verification in case of Apache Mina is handled by two providers bc and eddsa. KnownHostKeyVerifier is called for verification of remote server keys & checks if the keys belong to which algos. So only the eddsa and similar curves are supported by eddsa dependency, rest all including non-ec and other ec curves are supported by bouncy castle itself.
[https://github.com/apache/mina-sshd/blob/62e6bfef2666655a3939a80499c20b8a274da4b3/sshd-common/src/main/java/org/apache/sshd/common/config/keys/KeyUtils.java#L1065](code-mina)

Also, as this eddsa library is not fips compliant it'll be better to release two separate artifacts (two separate libraries) one fips compliant and one fips non compliant. One can have more algos supported one can only have fips compliant algos supported. @acurci-at-mulesoft Please provide your opinion

Reply:
As discussed in the meeting, we'll be removing this dependency from the pom, since it is not FIPS compliant.